### PR TITLE
Migrate stock_inventory_hierarchical to v8

### PR DESCRIPTION
**Pull request by clonedagain**
Migrate stock_inventory_hierarchical to v8
# Migration/adaptation to Odoo v8

Use `stock.inventory.location_id` (now it's standard)
Move the "generate inventories" wizard to stock_inventory_hierarchical (drop the field exhaustive)
Use v8 API
Drop the exception `HierarchicalInventoryException` in favor of the new standard exception `ValidationError`
Place the sub-inventories outside the notebook on the main inventory form (the notebook is hidden on draft inventories)
Update the translation template and translation into French.
# Improvements

When creating a hierarchy of inventories using the wizard with the "views only" feature, make sure it matches the view hierarchy. In v7 it would ignore views which were inside internal locations, but this is more common in v8 because a location can't be the source or destination of any stock moves.
Automatically open the sub-inventories when the main inventory is opened. This avoids having to compute useless lines for the main inventory and then re-computing the same lines when opening the sub-inventories
Automatically cancel (set to draft) the sub-inventories when the main inventory is canceled.
Add warnings when clicking the buttons "Start inventory" and "Cancel inventory". The two changes above are a bit invasive, so we must make sure users know what's going to happen.
Speed improvement: defer the computation of `parent_left`/`parent_right` to the end of the wizard.
Add documentation (cf README.rst)
# Coding standards

Clearly define (and strictly enforce) the consitency of the states of the inventories with their sub-inventories.
Rewrite the tests in API v8, improve the coverage. Drop the demo data in favor of a much larger test dataset, automatically generated during the tests.
Adopt the newest OCA standards for licences, README
Bump version to 2.0
Update the screen captures

### CODE REVIEW

**Comment by clonedagain** [comment]
@laetitia-gangloff this PR contains the migration of the first module (parent/child relationship). Can you please review it and maybe test it on copies of your customer's databases?

**Comment by clonedagain** [comment]
There seems to be a bad interaction with some other modules in the repo, we're investigating.

**Comment by clonedagain** [comment]
Problem fixed, ready for review.

File: stock_inventory_hierarchical/README.rst [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+Hierarchical structure for Inventory adjustments
+================================================
+
+This module adds a parent-child relationship between inventory adjustments, to
+help users manage complex warehouses.
+Using several inventories, you can distribute the counting to several persons
+and still keep a clear overview of global inventory's status.
+
+The main inventory will be for a general Location, and the sub-inventories
+should contain sub-Locations.
```
Comment (elicoidal) [inline stock_inventory_hierarchical/README.rst]: sub-locations

File: stock_inventory_hierarchical/README.rst [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+Odoo will make sure the status of the Inventory and it's Sub-Inventories are
```
Comment (elicoidal) [inline stock_inventory_hierarchical/README.rst]: inventory / sub-inventories

File: stock_inventory_hierarchical/README.rst [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+consistent:
+
+* sub-inventories will be prepared automatically when you prepare the main
+  inventory
+* you must finish the sub-inventories before you can finish the main inventory
+* canceling a sub-inventory is only possible is the parent inventories are
```
Comment (elicoidal) [inline stock_inventory_hierarchical/README.rst]: should not it be the other way around?
you can only cancel a parent inventory if all sub-inventories are cancelled?

File: stock_inventory_hierarchical/README.rst [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+.. image:: https://odoo-community.org/website/image/ir.attachment/5784_f2813bd/datas
+   :alt: Try me on Runbot
+   :target: https://runbot.odoo-community.org/runbot/153/8.0
+   
+Generating the inventories
+--------------------------
+
+The most convenient way to create a complete hierarchy of inventories is the
+wizard available in the menu
+"Warehouse > Inventory Control > Generate Inventory"
```
Comment (elicoidal) [inline stock_inventory_hierarchical/README.rst]: please use enumerated list

File: stock_inventory_hierarchical/README.rst [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+
+If that doesn't fit your warehouse, you can also create the inventories
+manually, simply indicating the parent for each one before you start them.
+
+Processing the inventories
+--------------------------
+Preparing the main inventory will automatically prepare the sub-inventories.
+All the sub-inventories will set to the same date as the main inventory.
```
Comment (elicoidal) [inline stock_inventory_hierarchical/README.rst]: All the sub-inventories will have the same date as the main inventory.

File: stock_inventory_hierarchical/__openerp__.py [hunk 6]
```diff
@@ -0,0 +1,19 @@
+# -*- coding: utf-8 -*-
+# © 2013-2016 Numérigraphe SARL
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+{
+    "name": "Hierarchical Inventory adjustments",
+    "summary": "Group several Inventory adjustments in a master inventory",
+    "version": "8.0.2.0.0",
+    "depends": ["stock"],
+    "author": u"Numérigraphe,Odoo Community Association (OCA)",
```
Comment (elicoidal) [inline stock_inventory_hierarchical/__openerp__.py:10]: I am always confused with the "u": is it really necessary? some says no, some yes...

File: stock_inventory_hierarchical/i18n/fr.po [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+#. module: stock_inventory_hierarchical
+#: field:stock.generate.inventory,create_date:0
+msgid "Created on"
+msgstr "Créé le"
+
+#. module: stock_inventory_hierarchical
+#: code:addons/stock_inventory_hierarchical/models/stock_inventory.py:90
+#, python-format
+msgid "Error: You can not create recursive inventories."
+msgstr "Erreur: Vous ne pouvez pas créer d'inventaire récursif"
```
Comment (elicoidal) [inline stock_inventory_hierarchical/i18n/fr.po]: period

File: stock_inventory_hierarchical_location/__init__.py [hunk 8]
```diff
@@ -0,0 +1,22 @@
+# -*- coding: utf-8 -*-
```
Comment (elicoidal) [inline stock_inventory_hierarchical_location/__init__.py:1]: Please simplify header

File: stock_inventory_hierarchical_location/__openerp__.py [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+#
+#    You should have received a copy of the GNU General Public License
+#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+##############################################################################
+
+{
+    "name": "Exhaustive and hierarchical Stock Inventories",
+    "summary": "Extra consistency checks",
+    "version": "7.0.1.1.0",
```
Comment (elicoidal) [inline stock_inventory_hierarchical_location/__openerp__.py:24]: This is a v7 module?

File: stock_inventory_hierarchical/__openerp__.py (hunk 6 above)
Comment (clonedagain) [inline stock_inventory_hierarchical/__openerp__.py:10]: It is a reasonable precaution because of the accent. Various applications read this file (odoo, the app store, the OCA bots...) and I don't expect all to take unicode for granted.

File: stock_inventory_hierarchical_location/__openerp__.py (hunk 9 above)
Comment (clonedagain) [inline stock_inventory_hierarchical_location/__openerp__.py:24]: Yes, but this is not the module we're migrating in this PR. This is a side-effect of bringing the history in the repo. We need to bring 3 modules's history into the repo together because they have lots of common commits.

If you'd like to help us make it clearer, can you please review this PR before? https://github.com/OCA/stock-logistics-warehouse/pull/137
This way only the last commit would be shown here.

File: stock_inventory_hierarchical/README.rst (hunk 3 above)
Comment (clonedagain) [inline stock_inventory_hierarchical/README.rst]: You're right, we kept this in the code but the "cancel" state is not reachable in the standard, so it does not deserve to be in the README.

File: stock_inventory_hierarchical_location/__init__.py (hunk 8 above)
Comment (clonedagain) [inline stock_inventory_hierarchical_location/__init__.py:1]: It will be done in another PR when we migrate this module (please cf. my reply regarding  `stock_inventory_hierarchical_location/__openerp__.py`)

File: stock_inventory_hierarchical/wizard/generate_inventory.py [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+                                  string='Location', required=True)
+    level = fields.Integer(string="Level", default=3,
+                           help="Maximum number of intermediate "
+                                "sub-inventories between the main inventory "
+                                "and the smallest sub-inventory.")
+    only_view = fields.Boolean(string='Only view', default=True,
+                               help="If set, only inventory "
+                                    "on view location can be created")
+
+    @api.multi
```
Comment (gurneyalex) [inline stock_inventory_hierarchical/wizard/generate_inventory.py]: @api.model

File: stock_inventory_hierarchical/wizard/generate_inventory.py [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+    def _default_location(self):
+        """Default stock location
+
+        @return: id of the stock location of the first warehouse of the
+        default company"""
+        company_id = self.env['res.company']._company_default_get(
+            'stock.warehouse')
+        warehouse_ids = self.env['stock.warehouse'].search(
+            [('company_id', '=', company_id)], limit=1)
+        return warehouse_ids and warehouse_ids.lot_stock_id.id or False
```
Comment (gurneyalex) [inline stock_inventory_hierarchical/wizard/generate_inventory.py]: `return warehouse_ids.lot_stock_id.id` (new api magic)

File: stock_inventory_hierarchical/wizard/generate_inventory.py [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+        # Create the sub-inventories
+        for _dummy in range(1, self.level):
+            inventories = fast_self._create_subinventory(
+                inventories, prefix_inv_name, self.only_view)
+        # Compute the parent_left/right
+        fast_self.env['stock.inventory']._parent_store_compute()
+
+        result = self.env['ir.model.data'].get_object_reference(
+            'stock', 'view_inventory_form')
+        view_id = result and result[1] or False
```
Comment (gurneyalex) [inline stock_inventory_hierarchical/wizard/generate_inventory.py]: result[1] if result else False

File: stock_inventory_location/__openerp__.py [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+   is final and you are now counting the goods.
+   In that status, no Stock Moves can be recorded in/out of the Inventory's
+   Locations.
+ - if the Location or some of it's children have not been entered in the
+   Inventory Lines, OpenERP warns you when you confirm the Inventory.
+ - only the Inventory's Location or its children can be entered in the
+   Inventory Lines.
+ - every good that is not in the Inventory Lines is considered lost, and gets
+   moved out of the stock when you confirm the Inventory.
+""",
```
Comment (gurneyalex) [inline stock_inventory_location/__openerp__.py:59]: can you move this out to README.rst?

**Comment by gurneyalex** [comment]
:+1:

File: stock_inventory_location/__openerp__.py (hunk 13 above)
Comment (clonedagain) [inline stock_inventory_location/__openerp__.py:59]: Will do in an upcoming PR targetting this module specifically. For now it is unported.

**Comment by clonedagain** [comment]
Rebased to clean up the diff, and corrected as @gurneyalex suggested.

**Comment by jgrandguillaume** [comment]
Hi,

Thanks for the contribs, I've just tested it and works fine. Great module !

:+1: 

Regards,

Joël

**Comment by clonedagain** [comment]
@lbellier and I just fixed a missing cache invalidation here, and added a test which demonstrates it. Reviews are still welcome.

**Comment by gurneyalex** [comment]
@clonedagain is this ready for merge or are you still working on it?

**Comment by clonedagain** [comment]
I think it's ready for merge. @lbellier and I ported the last related module yesterday 
so I don't expect significant changes now.

**Comment by lmignon** [comment]
:+1: Successfully used in production.

**Comment by clonedagain** [comment]
@lmignon are you using v8 in production already? there's been quite a bit of refactoring due to the standard having locations in inventories.

**Comment by lmignon** [comment]
@clonedagain Oups, we use the V7 version... too fast

**Comment by lmignon** [comment]
@clonedagain I've deleted my comments with the approval since it was for the V7.... I've read deeper the code and I'm really impressed by the new way inventories are created and prepared. Even if the code is quite complex and use knowingly some SQL queries, it stay readable. To complete my review I should do some functional tests but I've no time at this moment.

**Comment by clonedagain** [comment]
Can we merge this please?
